### PR TITLE
feat(collections): Add BinaryHeap

### PR DESCRIPTION
**Pull request by KyleJune**
Binary heaps are common data structures in other languages but JavaScript does not have a built in BinaryHeap class. In Rust, they have a binary heap in std/collections that can be found [here](https://doc.rust-lang.org/stable/std/collections/struct.BinaryHeap.html). In Java they have one too but call it [PriorityQueue](https://docs.oracle.com/javase/7/docs/api/java/util/PriorityQueue.html). I thought it would be handy if Deno had one too. I decided to go with the name BinaryHeap to match Rust's BinaryHeap.

Originally I implemented this in a third party module called [collections](https://deno.land/x/collections). In an issue for adding my [test_suite](https://deno.land/x/test_suite) module to std/testing, the possibility of adding my [mock](https://deno.land/x/mock) module came up too. The mock module makes use of a Red Black Tree that is part of my collections module and @bartlomieju said he would be open to adding some of my other structures to std/collections too.

https://github.com/denoland/deno_std/issues/1991#issuecomment-1061315245
> Correct, our implementation of RB tree is internal and can't be exposed. I'm also open to adding this module to `std/collections` as well as other structures you have there. They seem very handy +1

As a first step I thought it would be better to start small and just add BinaryHeap first before attempting to add my red black tree implementation.

Below are links to the deno docs for the 2 files added as part of this PR.

- [binary_heap.ts](https://doc.deno.land/https://raw.githubusercontent.com/KyleJune/deno_std/kj-add-binary-heap/collections/binary_heap.ts)
- [comparators.ts](https://doc.deno.land/https://raw.githubusercontent.com/KyleJune/deno_std/kj-add-binary-heap/collections/comparators.ts)

### CODE REVIEW

File: collections/binary_heap_test.ts
Comment (bartlomieju) [inline collections/binary_heap_test.ts:11]: These two don't need to be `export`ed

File: collections/comparators.ts
Comment (bartlomieju) [inline collections/comparators.ts:11]: I think this module is not needed; if it is used in `binary_heap.ts` we could just define those inline in that module.

**Review by bartlomieju — COMMENTED** [review]
I'm in favor of landing this PR. @kt3k any objections?

File: collections/binary_heap_test.ts
Comment (KyleJune) [reply collections/binary_heap_test.ts:11]: My bad. I originally had these in a different file called test_common.ts which I use in multiple test files of mine. I forgot to remove the export when I copied it into collections/binary_heap_test.ts. If I'm going to be using these objects again in my binary search tree and red black tree test coverage, should I inline them like I did here or create a test data folder for storing these classes used in multiple test cases?

File: collections/comparators.ts
Comment (KyleJune) [reply collections/comparators.ts:11]: When I add BSTree and RBTree in separate PRs, they will use these comparators too for determining how to sort the values added to the tree. That's why I put them in their own file. These functions work with `Array.prototype.sort` too. I'm fine with moving them too though. Having them at least somewhere is useful to avoid having to re-implement these functions whenever you want to change the default sort behavior of a data structure or sort an existing one.

Comment (KyleJune) [reply collections/comparators.ts:11]: Here is a usage example for making the BinaryHeap be a Min Heap instead of a Max Heap.

```ts
import { BinaryHeap, ascend } from "https://deno.land/std@0.129.0/collections/mod.ts";
const maxHeap = new BinaryHeap<number>(); // descend is default compare function.
const minHeap = new BinaryHeap<number>(ascend);
```

Comment (KyleJune) [reply collections/comparators.ts:11]: Another option would be to keep them in their own separate file but also update binary_heap.ts to re-export them for easy access, that way you could import them all from binary_heap.ts instead of importing the entire mod.ts or having 2 separate import lines.

**Comment by kt3k** [comment]
@bartlomieju I'm in favor of including these kinds of data structures in general. Should be useful for algorithmic works. Will review in more details soon

File: collections/binary_heap_test.ts
Comment (kt3k) [reply collections/binary_heap_test.ts:11]: We loosely have the rule in deno_std that if the filename started with `_`, then that file is considered private. So maybe we can define `collections/_test_utils.ts` if we need some common items only used for testing.

File: collections/comparators.ts
Comment (kt3k) [reply collections/comparators.ts:11]: > Another option would be to keep them in their own separate file but also update binary_heap.ts to re-export them for easy access

This sounds better option to me. And also I think it's better to prefix this module as `_comparators.ts` to prevent the direct dependency to this file.

File: collections/binary_heap_test.ts
Comment (KyleJune) [reply collections/binary_heap_test.ts]: Done, I've moved those 2 classes to `collections/_test_utils.ts`.

File: collections/comparators.ts
Comment (KyleJune) [reply collections/comparators.ts]: I added an underscore prefix to the `comparators.ts` and `comparators_test.ts` as requested and updated `binary_heap.ts` to export them.

I've also updated the README.md file to have a section for BinaryHeap, showing the efficiency of the functions. I included an example of both min and max heaps along with showing how the comparator functions can be used to compare properties of the values in a BinaryHeap.

**Review by bartlomieju — APPROVED** [review]
LGTM, thank you for this @KyleJune. I'll wait for @kt3k review before landing.

**Review by kt3k — APPROVED** [review]
@KyleJune LGTM. Great!

**Pull request merged**
